### ISSUE TITLE
[Datepicker] support intl locales (non-gregorian calendars)

### ISSUE THREAD

**Issue by sijad** [uber/baseweb#2594]
## Feature description

adding intl support to Datepicker

**Comment by gergelyke** [comment]
thank you for opening this @sijad - do you think you could help us implement this?

**Comment by sijad** [comment]
sure, I've cheked both of those examples and I think this can be implimented using `intl.DateTimeFormat`

material-ui also were using intl before but now it's using date-io/jalaali (which uses moment internally). I'm not sure why they did droped intl?

**Comment by gergelyke** [comment]
> material-ui also were using intl before but now it's using date-io/jalaali (which uses moment internally). I'm not sure why they did droped intl?

yeah, I am not sure - you may be able to find some PRs and discussion around it

**Comment by sijad** [comment]
currently baseui uses `date-fns.format` to format dates, which does not support non-gregorian calendars. changing it to use `intl.DateTimeFormat` would be a breaking change, any thoughts?

**Comment by gergelyke** [comment]
> currently baseui uses `date-fns.format` to format dates, which does not support non-gregorian calendars. changing it to use `intl.DateTimeFormat` would be a breaking change, any thoughts?
> 
> edit: it's a breaking change because of `formatString` and `locale` props.

we could add a property to make this opt-in, so it wouldn't be a breaking change - would that work?

**Comment by sijad** [comment]
I've checked zendesk garden implementation and find out it's not a perfect solution because all of its calculations are based on Gregorian date and only show formatted values in UI and I think it's confusing for users.
for example number of days in a month might be different in Jalali or Hijri and Gregorian.

it's be possible to implement this correctly if it was possible to pass alternative date functions (e.g. `getStartOfWeek`, `getStartOfMonth`, ...) in run time via props (similar to `material-ui-pickers`). is it a good idea?

**Comment by gergelyke** [comment]
I think that makes sense - @nadiia do you think it should be done through top level props, or the overrides interface?
